README.md
# cacheController

a cache controller for cache data from request, can be auto read when a request have finished. and auto remove the least use cache.

Get started with the documentation, or play with the examples (see how to run them below).
# Features

缓存控制插件
用于和请求混合使用以达到缓存请求的作用
# Install && Use
```shell
npm i cachecontroller

```
```javascript

import CacheController from 'cachecontroller'
const requestCacheController = new CacheController({ maxSize: 50, maxUseTime: 3 })

```



# Example
 1. 配合请求拦截器使用
```javascript
// axios.js

import CacheController from 'cachecontroller'
const requestCacheController = new CacheController({ maxSize: 50, maxUseTime: 3 })


// 创建axios实例
const service = axios.create({
    baseURL: '/api',
    timeout: 15000
})


const request = async (config) =>{
    //  可自行实现
        const generateDataStr = config => {
            const { method, data, params } = config
            let dataStr = ''
            if (method === 'get' && params) {
                dataStr = typeof params !== 'string' ? JSON.stringify(params) : params
            }
            if (method === 'post' && data) {
                dataStr = typeof data !== 'string' ? JSON.stringify(data) : data
            }
            return dataStr
        }
        const dataStr = generateDataStr(config)
        const hash = `${method}${url}${dataStr}${unique ? Math.random() : ''}`
        return new Promise((resolve, reject) => { 
        if (requestCacheController.haveCache(hash)) {
            return resolve(requestCacheController.getCache(hash))   
            }else{
            const data = await service.request(config)
            if(config.cache){
                requestCacheController.setCache(hash, { data, size: 1 })
            }
            return data
        }
    })
}

// api.js

import request from 'axios.js'

export function addAddress (params) {
    return request({
        url:'/address',
        params,
        cache:true
    })
}

```
2. 配合缓存生成器使用

```javascript 
import cacheController from 'cachecontroller'
const requestCacheController = new cacheController({ maxSize: 50, maxUseTime: 3 })

const request = (params)=>{    
   return fetch(`https://api.apiopen.top/api/sentences?page=${params.page}&size=${params.size}`)
    .then((res)=> res.json())
    .then((json)=> json)
}
const params = {
    page:1,
    size:2
}
const api = requestCacheController.generateRequest(`sentences`,request)
    api(params).then(res=>{
        if(res.code===200){
        // do sth
        }
    }).catch(e=>{
        // console.log(e)
    })

```

## Params 实例参数说明

|  参数   | 默认值  | 说明  |   类型|
|  ----  | ----  |----|----|
| maxSize|10 |储存的数据大小当装满了会尝试清空最早使用的数据腾出空间 | number |
| maxUseTime  | 3 | 最大使用次数，当从缓存中读取次数超过该数字缓存将被移除。| number

## Methods 方法说明

|  方法名   | 类型  | 说明  |  
|  ---- | -----  |----|
| setCache | (key: string, blob: { size : number, data : any }) => void   | 存入缓存，key值唯一，当重复出现会覆盖。bolb size 必传且必必须是数字  |
| haveCache  | (key: string) => boolean | 返回缓存中是否含有对应key的缓存不计入读取次数 |
| getCache  | (key: string) => any or null | 返回缓存中有对应key的缓存 |  (any or null)
| removeEarliestCache | () => void | 删除最早使用的key对应的缓存 | void
| deleteCache | (key: string) => void | 删除key对应的缓存 |
|generateRequest| (key: string, ()=> Promise) => void | 为一个请求方法生成一个自动缓存的请求方法 |

## Development Setup  开发启动

```shell

# install deps

npm install

# build dist files

npm run build

# serve examples at localhost:8080

npm run serve

# lint & run all tests

npm test

```

## Todo list

- [X] 实现生成器，对请求函数进行封装，返回一个可执行的promise 实现动态生成缓存请求方法。
- [ ] 实现缓存持久化储存hook，提供客户端持久化能力。

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/emjio/cacheController/compare/v1.1.0...v1.2.0) (2023-01-31)


### Features

* 代码格式化 ([697bc72](https://github.com/emjio/cacheController/commit/697bc7232ada81ce303b31d76ec0c67ed6b76c35))
* 完善testcase 配置。新增请求缓存生成器逻辑 ([586064d](https://github.com/emjio/cacheController/commit/586064df50cd0a6d7f8c1fff42da96a2a25aee2b))
* 属性值合法校验 ([1a94b16](https://github.com/emjio/cacheController/commit/1a94b16b7f09b0a229223551daa4a093777131e7))
* 新增eslint ([731c5e8](https://github.com/emjio/cacheController/commit/731c5e8f552bee4305599e37002b02d70177b3b7))
* 新增example ([e9f57d3](https://github.com/emjio/cacheController/commit/e9f57d39d57842f875d066d4a50c4219222fed0c))
* 新增类型声明，完善说明文档 ([d679234](https://github.com/emjio/cacheController/commit/d67923480c6a98623be9fb149bf96c496a5b20df))
* 新增请求缓存函数生成器 ([50874aa](https://github.com/emjio/cacheController/commit/50874aa4c483b6c9e96ce0e3b19ef81651c3aeaf))
* 新增请求缓存函数生成器 ([7f5d477](https://github.com/emjio/cacheController/commit/7f5d4770268cecb306074d73c791f3b7933776ea))
* 重构rollup 配置。 ([f4a76c8](https://github.com/emjio/cacheController/commit/f4a76c87aea83d79e484f496176345aff080a65d))


### Bug Fixes

* typo fix ([687852e](https://github.com/emjio/cacheController/commit/687852e5456a1181ef73f738046b98d0583e666d))

## 1.1.0 (2023-01-04)


### Features

* init ([66e362d](https://github.com/emjio/cacheController/commit/66e362d54d2fd8b011e34049adf4f116e3a4fafc))
* 新增版本控制 ([b44a26b](https://github.com/emjio/cacheController/commit/b44a26be7b2d9c6b8e463776d48926ed1af7bd50))
* 新增版本控制 ([518cc3c](https://github.com/emjio/cacheController/commit/518cc3ca8a50ea35c24ef6106d2af03d6d894806))


example/basic/app.js
import cacheController from 'cachecontroller'
import axios from 'axios'
const requestCacheController = new cacheController({ maxSize: 50, maxUseTime: 3 })




// 创建axios实例
const service = axios.create({
    baseURL: 'https://api.apiopen.top/api',
    timeout: 15000
})

const generateDataStr = config => {
            const { method, data, params } = config
            let dataStr = ''
            if (method === 'get' && params) {
                dataStr = typeof params !== 'string' ? JSON.stringify(params) : params
            }
            if (method === 'post' && data) {
                dataStr = typeof data !== 'string' ? JSON.stringify(data) : data
            }
            return dataStr
        }

const request = async (config) =>{
         const { method, data, params,url } = config
        const dataStr = generateDataStr(config)
        const hash = `${method}${url}${dataStr}`
        return new Promise(async(resolve, reject) => { 
        if (requestCacheController.haveCache(hash)) {
            console.log('读取到缓存')
            return resolve(requestCacheController.getCache(hash))   
            }else{
                console.log('从网络获取')
            const data = await service.request(config)
            if(config.cache){
                requestCacheController.setCache(hash, { data:data.data, size: 1 })
            }
            return resolve(data.data)
        }
    })
}

const sendbtn = document.querySelector('#sendRequest')
const resultContent = document.querySelector('#result')

const params = {
    page:1,
    size:2
}

 function getSentences(params) {
    return request({
        url:'/sentences',
        params,
        method:'get',
        cache:true
    })
}
sendbtn.addEventListener('click',()=>{
    getSentences(params).then(res=>{
        if(res.code===200){
            resultContent.innerText = `${res.result.name} ---${res.result.from}`
        }
    }) 
})